Retries metric

tmasternak opened the pull request:
## Description
This PR introduces retries metrics. It does so by processing `Occurrences` message published by `NSB.Metrics`. In introduces following changes:
 * Deserializer for `Occurrences` messages
 * Handler for `Occurrences` messages that records data in `RetriesStore` based on `IntervalStore` 
 * Renaming `TimingStore` to `IntervalStore`

mikeminutillo commented on src/ServiceControl.Monitoring/Metrics/Raw/OccurrenceSerializerDefinition.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        public object[] Deserialize(Stream stream, IList<Type> messageTypes = null)
+        {
+            long baseTicks = 0;
```
I'm not sure about this being captured in an enclosure. Wondering if we should create a `state` struct and pass that around. It can include the creation of the message pool and set up the base ticks on the starter and then get passed in to each subsequent call.

mikeminutillo replied:
It turned out to be simpler to just drop the timestamp management into the base class. Both current implementation manage timestamps the same way. 

https://github.com/Particular/ServiceControl.Monitoring/pull/38

mikeminutillo commented on src/ServiceControl.Monitoring/Retries/Retries.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace ServiceControl.Monitoring
+{
+    public class Retries
```
What is this?

tmasternak replied:
I have no idea :P

mikeminutillo commented on src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs:
```diff
@@ -12,9 +12,9 @@ public class MonitoredEndpointsModule : ApiModule
         /// Initializes the metric API module.
         /// </summary>
-        public MonitoredEndpointsModule(ProcessingTimeStore processingTimeStore, CriticalTimeStore criticalTimeStore, QueueLengthDataStore queueLengthDataStore)
+        public MonitoredEndpointsModule(ProcessingTimeStore processingTimeStore, CriticalTimeStore criticalTimeStore, QueueLengthDataStore queueLengthDataStore, RetriesStore retriesStore)
```
You should just be able to replace this with a dependency on `TimingsAggregator` and Autofac will put everything together.

WilliamBZA commented on src/ServiceControl.Monitoring/Retries/OccurrenceHandler.cs:
```diff
@@ -0,0 +1,41 @@
+                        store.Store(endpointInstanceId, message.Entries);
+                        break;
+                    default:
+                        throw new UnknownOccurrenceMessageType(messageType);
```
Should we have a custom retry policy so these messages get forwarded straight to the error queue instead of going through FLR+SLR?

WilliamBZA replied:
Or perhaps even discarded?

WilliamBZA commented on src/ServiceControl.Monitoring/Metrics/Raw/RawMessage.cs:
```diff
@@ -0,0 +1,50 @@
+            Array.Clear(Entries, 0, MaxEntries);
+        }
+
+        public Entry[] Entries = new Entry[MaxEntries];
```
Much better name :+1:

WilliamBZA commented on src/ServiceControl.Monitoring/Metrics/Raw/OccurrenceSerializerDefinition.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        public object[] Deserialize(Stream stream, IList<Type> messageTypes = null)
+        {
+            return DeserializeRawMessage(stream, (t, r, m) => m.TryRecord(t));
```
My OCD here, but can we give these better names than `t`, `r`, and `m`? Especially since the func's types aren't named.

Scooletz replied:
Is there a lambda allocation in here? If yes, could we extract the creation of the delegate and pass an already constructed one?

tmasternak replied:
Solved by Store method in base class.

WilliamBZA commented on src/ServiceControl.Monitoring/Metrics/Raw/LongValueOccurenceSerializerDefinition.cs:
```diff
@@ -15,58 +15,17 @@ class LongValueOccurrenceSerializerDefinition : SerializationDefinition
-
-                if (count == 0)
+            return DeserializeRawMessage(stream,
+                (t, r, m) =>
```
My OCD here, but can we give these better names than `t`, `r`, and `m`? Especially since the func's types aren't named.

Scooletz replied:
Is there a lambda allocation in here? If yes, could we extract the creation of the delegate and pass an already constructed one?

tmasternak replied:
This is solved by abstract `Store` method in base class.

tmasternak replied:
Solved by `Store` method in base class.

Scooletz commented on src/ServiceControl.Monitoring/Metrics/Raw/RawMessageSerializer.cs:
```diff
@@ -0,0 +1,59 @@
+                {
+                    if (i % RawMessage.MaxEntries == 0)
+                    {
+                        message = RawMessage.Pool<T>.Default.Lease();
```
Tricky :)

Scooletz commented on src/ServiceControl.Monitoring/Metrics/Raw/RawMessageSerializer.cs:
```diff
@@ -0,0 +1,59 @@
+            throw new NotImplementedException();
+        }
+
+        protected object[] DeserializeRawMessage(Stream stream, Func<long, BinaryReader, T, bool> store)
```
`store` could be an `abstract` method of this type (`RawMessageSerializer` is used only as a base class) removing the need of passing the storing delegate.

tmasternak replied:
Done.

Scooletz commented on src/ServiceControl.Monitoring/Timings/IntervalsStore.cs:
```diff
@@ -3,33 +3,35 @@
     using Metrics.Raw;
 
-    public abstract class TimingsStore
+    public abstract class IntervalsStore : IKnowAboutEndpoints
```
I'm not a native English speaker but this name is somewhat strange `IntervalsStore`. Shouldn't it be `IntervalStore`?

tmasternak replied:
I would kept the plural to follow `TimingsStore` convention.

Scooletz commented on src/ServiceControl.Monitoring/Retries/RetriesStore.cs:
```diff
@@ -0,0 +1,46 @@
+    using System.Linq;
+    using Infrastructure;
+    using Timings;
+    public class RetriesStore : IntervalsStore, IProvideEndpointMonitoringData, IProvideEndpointInstanceMonitoringData
```
I'm not a native English speaker but this name is somewhat strange `RetriesStore`. Shouldn't it be `RetryStore `?

tmasternak replied:
Same as [above](https://github.com/Particular/ServiceControl.Monitoring/pull/37#discussion_r129402725).

Scooletz commented on src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs:
```diff
@@ -111,6 +64,7 @@ public class MonitoredEndpoint
         public string[] EndpointInstanceIds { get; set; }
         public MonitoredEndpointValues ProcessingTime { get; set; }
         public MonitoredEndpointValues CriticalTime { get; set; }
+        public MonitoredEndpointValues Retries { get; set; }
```
So we abstract `IProvideEndpointInstanceMonitoringData` and `IProvideEndpointMonitoringData` implementations but still keep the properties in here, "knowing" what are `Retries`, `ProcessingTime`, `CriticalTime`. It seems incoherent to me. A dictionary-like approach, with implementations of `IProvide...` adding keys, would be better maybe?

tmasternak replied:
I propose to make this a separate PR in which I will try to make `MonitoredEndpointValue` more generic.

mikeminutillo replied:
Yeah. This is certainly the intent. That change needs to be done in conjunction with the UI though as it will affect the UI.

Scooletz commented on src/ServiceControl.Monitoring/Timings/CriticalTimeStore.cs:
```diff
@@ -1,7 +1,29 @@
+
+            foreach (var item in data)
+            {
+                item.CriticalTime = AggregateTimings(snapshot[item.Name].ToList());
```
Will this lookup `snapshot` always include `item.Name`?

tmasternak replied:
No but in such case an empty list will get created.

Scooletz commented on src/ServiceControl.Monitoring/Timings/ProcessingTimeStore.cs:
```diff
@@ -1,7 +1,28 @@
+
+            foreach (var item in data)
+            {
+                item.ProcessingTime = AggregateTimings(snapshot[item.Name].ToList());
```
Will this lookup `snapshot` always include `item.Name`?

tmasternak replied:
No but in such case an empty list will get created.

tmasternak commented:
@Scooletz I've addressed your comments. I think that performance tests and making `MonitoredEndpoint` class more generic should be addressed in separate PRs.

Scooletz commented:
LGTM @tmasternak

mikeminutillo commented on src/ServiceControl.Monitoring/Timings/ProcessingTimeStore.cs:
```diff
@@ -1,7 +1,28 @@
+
+        public void FillIn(MonitoredEndpointInstance[] data, DateTime now)
+        {
+            var snapshot = GetIntervals(now).ToLookup(k => $"{k.Id.EndpointName}-{k.Id.InstanceId}");
```
If the view model included a proper EndpointInstanceId then we could just lookup based on that.

tmasternak commented:
@mikeminutillo can we merge this?
